Ensure parallel map `Closed` error propagation

johnhungerford opened the pull request:
<!--
PRs require an approval from any of the core contributors, other than the PR author.

Include this header if applicable:
Fixes #issue1, #issue2, ...
-->

Fixes #1387 

### Problem
<!--
Explain here the context, and why you're making this change. What is the problem you're trying to solve?
-->

For concurrent stream transformations using `Channel` and/or `Meter`, there is a bug (#1387) where if the transformation effect `f` fails with `Abort[Closed]`, that failure is interpreted as end-of-stream.

### Solution
<!--
Describe your solution. Focus on helping reviewers understand your technical approach and implementation decisions.
-->

Added `Abort.withMask` which allows you to mask an `Abort` failure type that may or may not be within a generic type intersection to protect it from being handled accidentally.

### Notes
<!--
Add any important additional information as bullet points, such as:
- Implementation details reviewers should know about
- Open questions and concerns
- Limitations
-->

A couple significant changes to make this work:
1. `ConcreteTag` derivation now does an implicit search for `createSingle` so that the tag for a union can be derived from tags of single types
2. `Fiber.init` and other `Async` methods do not use `Reducible` to eliminate `Abort[Nothing]`. Instead, the `S` type parameter in `Fiber` is constrained to being a subtype of `Abort[Nothing]` to reflect the coupling of `Fiber` to failures.

johnhungerford committed:
Added Abort.withMask; updated ConcreteTag to derive from individual types

johnhungerford committed:
Added & using Abort.withMask

johnhungerford committed:
Fix readme

johnhungerford committed:
Merge branch 'main' into map-par-closed-bug

johnhungerford committed:
Added comment; removed redundant Scope.ensure from Actor

johnhungerford committed:
Merge branch 'map-par-closed-bug' of github.com:johnhungerford/kyo into map-par-closed-bug

fwbrasil commented on README.md:
```diff
@@ -1900,26 +1900,26 @@ val a: Unit < Sync =
 // Recurring task with a delay between
 // executions
-val b: Fiber[Unit, Any] < Sync =
+val b: Fiber[Unit, Abort[Nothing]] < Sync =
```
I think It'd be important to not track `Abort[Nothing]` here. Sorry, I think I indicated that it'd be fine to remove `Reducible` from `Fiber.init` but I thought this would still be `Any` since the underlying impl is `IOPromiseBase[Any, A < (Async & S)]` but I'm not sure it's possible.

johnhungerford replied:
I think it may still be possible to get rid of `Reducible` without tracking `Abort[Nothing]`. I'll give it a try

fwbrasil commented on kyo-data/shared/src/main/scala/kyo/internal/SafeClassTagMacro.scala:
```diff
@@ -50,45 +50,50 @@ private[kyo] object SafeClassTagMacro:
                             case _             => Chunk(tpe)
```
We need to rename from `SafeClassTagMacro` to `ConcreteTagMacro`

fwbrasil commented on kyo-data/shared/src/main/scala/kyo/internal/SafeClassTagMacro.scala:
```diff
@@ -50,45 +50,50 @@ private[kyo] object SafeClassTagMacro:
-                            case x: String  => Expr(x)
-                            case x          => report.errorAndAbort(s"Unsupported literal type: $x")
-                    '{ LiteralTag($value) }
+            Implicits.search(TypeRepr.of[[A] =>> ConcreteTag[A]].appliedTo(tpe)) match
```
The macro change looks ok to me, it should have been this way from the beginning. What do you think about separating this change in a new PR and adding tests for it?

fwbrasil commented on kyo-prelude/shared/src/main/scala/kyo/Abort.scala:
```diff
@@ -603,4 +603,61 @@ object Abort:
 
     end literal
 
+    sealed class CanMaskAbort
```
I can't see the purpose of this class. Maybe `MaskAbort .apply` should take it to ensure the masking is scoped?

johnhungerford replied:
`MaskAbort.apply` does take it. The signature starts with: `CanMaskAbort ?=> ...`

johnhungerford replied:
Oh good point, `MaskAbort.apply` should take it -- you're right

johnhungerford replied:
Added

fwbrasil commented on kyo-prelude/shared/src/main/scala/kyo/Abort.scala:
```diff
@@ -603,4 +603,61 @@ object Abort:
+          */
+        def apply[A, S](effect: A < (Abort[EG] & S)): A < (Abort[EG | Masked] & S)
+
+        given tag: ConcreteTag[Masked]
```
it seems this doesn't need to be in the abstract class?

johnhungerford replied:
I couldn't make it work without it here. If you need to call `Abort.run[mask.Masked]` within a `withMask` effect, this is the only way I can figure out to make the `ConcreteTag` available to the compiler.

fwbrasil commented on kyo-prelude/shared/src/main/scala/kyo/Abort.scala:
```diff
@@ -603,4 +603,61 @@ object Abort:
+            def apply[A, S](effect: A < (Abort[EG] & S)): A < (Abort[EG | Masked] & S) =
+                Abort.runPartial[E](effect).map:
+                    case Result.Success(a) => a
+                    case Result.Failure(e) => Abort.fail(Wr(e))
```
Nice insight :) I'm wondering if we could have a more generic version of this mechanism since it's essentially temporarily swapping the tag. I couldn't test it and there's some type safety challenges but I wonder if this could be a good path:

```scala
sealed trait Mask[E] extends ArrowEffect[Const[Any], Const[Any]]

object Mask:

    def apply[E <: ArrowEffect[?, ?]](using Frame)[A, S](v: A < (E & S))(using eTag: Tag[E], maskTag: Tag[Mask[E]]): A < (Mask[E] & S) =
        ArrowEffect.handle(eTag, v) {
            [C] => (input, cont) => ArrowEffect.suspendWith(maskTag, input)(cont)
        }

    def run[E <: ArrowEffect[?, ?]](using Frame)[A, S](v: A < (Mask[E] & S))(using eTag: Tag[E], maskTag: Tag[Mask[E]]): A < (E & S) =
        ArrowEffect.handle(maskTag, v) {
            [C] => (input: Any, cont) => ArrowEffect.suspendWith(eTag, input)(cont)
        }
```

It seems it should work to prevent `E` suspensions from being handled until `Mask.run`. Ah, but there's also the challenge that `Abort` still uses erased tags, though.

johnhungerford replied:
I tried something like this too, but I actually need to preserve the `Abort` so that I can do error handling. So I can't mask `Abort[E]` as `Mask[Abort[E]]`, but it would have to be `Abort[E]` -> `Abort[Mask[E]]`. That's why we need a specialized approach for abort.

johnhungerford committed:
Simplified withMask; applying to stream.emit as well

johnhungerford committed:
Reverted Fiber[A, Abort[Nothing]] to Fiber[A, Any]

johnhungerford committed:
Fixed readme

johnhungerford committed:
Merge remote-tracking branch 'upstream/main' into map-par-closed-bug

johnhungerford committed:
Merge branch 'main' into map-par-closed-bug

johnhungerford committed:
Merge branch 'main' into map-par-closed-bug

johnhungerford committed:
Removed redundant mailbox finalizer

johnhungerford committed:
Merge branch 'map-par-closed-bug' of github.com:johnhungerford/kyo into map-par-closed-bug

johnhungerford committed:
Formatted

johnhungerford committed:
Using sync.ensure again

johnhungerford committed:
Merge branch 'main' into map-par-closed-bug

hearnadam commented on kyo-core/shared/src/main/scala/kyo/Fiber.scala:
```diff
@@ -91,14 +90,14 @@ object Fiber:
       *   A Fiber that is immediately completed with the provided Result
       */
-    def fromResult[E, A, S](result: Result[E, A < S])(using reduce: Reducible[Abort[E]]): Fiber[A, reduce.SReduced & S] =
+    def fromResult[E, A, S](result: Result[E, A < S]): Fiber[A, Abort[E] & S] =
```
Does this need a frame?

fwbrasil commented:
@copilot resolve the merge conflicts in this pull request

ahoy-jon committed:
merge: main into map-par-closed-bug preserving Abort.withMask fix

Resolved conflicts in StreamCoreExtensions.scala by keeping the branch's
mask.Masked pattern across all four mapPar* methods. The mask makes user
errors (mask.Masked) and channel-close errors (Closed) statically disjoint,
which obviates main's AtomicRef[Maybe[Either[Throwable, Any]]] workaround.

Adopted main's Fiber.Unsafe.fromResult API in Channel/Fiber/Queue and added
.reduced in Exchange to bridge the Reducible-driven Fiber[A, Abort[Nothing]]
return with the branch's Fiber[A, Any] field type.

ahoy-jon commented:
@fwbrasil fixed the conflict with main
